### PR TITLE
8328648: Remove applet usage from JFileChooser tests bug4150029

### PR DESCRIPTION
**Pull request by alisenchung**
Removing applet usage from manual JFileChooser tests

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8328648](https://bugs.openjdk.org/browse/JDK-8328648): Remove applet usage from JFileChooser tests bug4150029 (**Bug** - P4)

### Reviewers
 * [Damon Nguyen](https://openjdk.org/census#dnguyen) (@DamonGuy - Committer) ⚠️ Review applies to [ed8ed220](https://git.openjdk.org/jdk/pull/18415/files/ed8ed2204cb25be8a920851c1da9b2f247b65f94)
 * [Abhishek Kumar](https://openjdk.org/census#abhiscxk) (@kumarabhi006 - Committer) ⚠️ Review applies to [ed8ed220](https://git.openjdk.org/jdk/pull/18415/files/ed8ed2204cb25be8a920851c1da9b2f247b65f94)
 * [Harshitha Onkar](https://openjdk.org/census#honkar) (@honkar-jdk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18415/head:pull/18415` \
`$ git checkout pull/18415`

Update a local copy of the PR: \
`$ git checkout pull/18415` \
`$ git pull https://git.openjdk.org/jdk.git pull/18415/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18415`

View PR using the GUI difftool: \
`$ git pr show -t 18415`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18415.diff">https://git.openjdk.org/jdk/pull/18415.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18415#issuecomment-2010825762)

### CODE REVIEW

**Comment by kumarabhi006** [comment]
This test can be converted to automated. Similar issue was resolved for GTK LAF. You can refer https://github.com/openjdk/jdk/pull/11291.

File: test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java [hunk 1]
```diff
@@ -21,73 +21,129 @@
...
-import javax.swing.*;
-import java.io.File;
-import java.io.IOException;
+/*
+ * @test
+ * @bug 4150029 8006087
+ * @summary BackSpace keyboard button does not lead to parent directory
+ * @library /test/lib /java/awt/regtesthelpers
+ * @build jdk.test.lib.Platform PassFailJFrame
+ * @run main/manual bug4150029
```
Comment (kumarabhi006) [inline test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java:43]: Not a manual test anymore, please remove manual test related tags and library.

File: test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java [hunk 2]
```diff
@@ -21,73 +21,129 @@
...
+            System.out.println("Created directory: " + testDir);
+
+            subDir = new File(testDir, "subDir");
+            subDir.mkdir();
+            subDir.deleteOnExit();
+            System.out.println("Created sub-directory: " + subDir);
+
+            fileChooser = new JFileChooser(subDir);
+
+//            fileChooser.setFileSelectionMode(JFileChooser.DIRECTORIES_ONLY);
```
Comment (kumarabhi006) [inline test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java:88]: Not required to keep commented code.

File: test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java [hunk 3]
```diff
@@ -21,73 +21,129 @@
...
+
+            SwingUtilities.invokeAndWait(() -> {
+                frame = new JFrame("Backspace Shortcut for Directory Navigation Test");
```
Comment (kumarabhi006) [inline test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java:91]: Block of code can be moved to separate method to create and show UI elements.

File: test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java [hunk 4]
```diff
@@ -21,73 +21,129 @@
...
-        File subDir = new File(testDir, "subDir");
-
-        subDir.mkdir();
-
-        System.out.println("Created directory: " + testDir);
-        System.out.println("Created sub-directory: " + subDir);
-
-        JFileChooser fileChooser = new JFileChooser(testDir);
+    private static void doTesting() {
+        Point p = frame.getLocationOnScreen();
```
Comment (kumarabhi006) [inline test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java:113]: should be on EDT.

File: test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java [hunk 5]
```diff
@@ -21,73 +21,129 @@
...
+        robot.mouseMove(p.x + 200, p.y + 200);
+        robot.mousePress(InputEvent.BUTTON1_MASK);
```
Comment (kumarabhi006) [inline test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java:115]:
```suggestion
        robot.mousePress(InputEvent.BUTTON1_DOWN_MASK);
```

should we add `robot.mouseRelease(InputEvent.BUTTON1_DOWN_MASK);` too ?

File: test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java [hunk 6]
```diff
@@ -21,73 +21,129 @@
...
+                break;
+            }
         }
 
-        try {
-            subDir.delete();
-            testDir.delete();
-        } catch (SecurityException e) {
-            e.printStackTrace();
+        if (passed_1 && passed_2) {
```
Comment (kumarabhi006) [inline test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java:136]: I think no need to print "Test Passed", can check for not passed and throw the runtime exception inside if block.

File: test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java [hunk 7]
```diff
@@ -21,73 +21,129 @@
...
+            fileChooser = new JFileChooser(subDir);
+
+//            fileChooser.setFileSelectionMode(JFileChooser.DIRECTORIES_ONLY);
+
+            SwingUtilities.invokeAndWait(() -> {
+                frame = new JFrame("Backspace Shortcut for Directory Navigation Test");
+                frame.getContentPane().setLayout(new BorderLayout());
+                fileChooser = new JFileChooser(subDir);
+                fileChooser.setControlButtonsAreShown(false);
+                frame.getContentPane().add(fileChooser, BorderLayout.CENTER);
```
Comment (DamonGuy) [inline test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java:95]:
```suggestion
                frame.add(fileChooser, BorderLayout.CENTER);
```

It was brought to my attention in one of my PRs that `getContentPane()` is obsolete as of JDK 1.5. They referenced  https://github.com/openjdk/jdk/pull/18390#discussion_r1533017080.

File: test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java [hunk 8]
```diff
@@ -21,73 +21,133 @@
...
-        System.out.println("Created sub-directory: " + subDir);
+    private static void doTesting() throws Exception {
+        SwingUtilities.invokeAndWait(() -> {
+            Point p = frame.getLocationOnScreen();
 
-        JFileChooser fileChooser = new JFileChooser(testDir);
+        });
+        robot.mouseMove(p.x + 200, p.y + 200);
+        robot.mousePress(InputEvent.BUTTON1_DOWN_MASK);
+        robot.mouseRelease(InputEvent.BUTTON1_DOWN_MASK);
```
Comment (DamonGuy) [inline test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java:121]: `p.x` and `p.y` isn't in scope I believe. You can move `Point p` outside the method, make it static and volatile, then it should be fine. Maybe remove the extra newline in this `SwingUtilities.invokeAndWait()` block as well.

**Review by DamonGuy — CHANGES_REQUESTED** [review]

File: test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java [hunk 9]
```diff
@@ -21,73 +21,133 @@
...
+        if (prevDir != crntDir) {
+            passed_1 = true;
+        }
 
-            e.printStackTrace();
+        // check if backspace key changes directory at root level
+        while (!fileChooser.getFileSystemView().isFileSystemRoot(prevDir)) {
+            clickBackSpace();
+            if (prevDir == crntDir) {
+                passed_2 = true;
```
Comment (kumarabhi006) [inline test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java:137]: This condition is not satisfied and hence the test failed in win11 machine. In linux machine, once the current directory is at `root` then pressing the backspace key doesn't have any effect.
But looks like it is different in windows and mac. Checked in windows native explorer, once it reached to `home` folder, it toggles between `home` and `Desktop` on backspace key press. The condition checked here will never satisfy and `passed_2` variable remains false.

So, either you can check the condition differently or may remove this check for windows and mac.

**Review by DamonGuy — COMMENTED** [review]
Looks like the test needs a headful tag as well. Was this ran via CI?

File: test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java [hunk 10]
```diff
@@ -21,73 +21,122 @@
...
-        } catch (IOException e) {
-            res = false;
+        boolean passed = false;
+        robot.waitForIdle();
 
-            e.printStackTrace();
+        // check backspace key at subDir level
+        clickBackSpace();
+        if (prevDir != crntDir) {
+            passed = true;
```
Comment (kumarabhi006) [inline test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java:128]: `passed` variable looks redundant now. You can directly evaluate compare the `prevDir` and `crntDir`, if they are same then test should fail.

File: test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java [hunk 11]
```diff
@@ -21,73 +21,119 @@
...
+ */
 
-public class bug4150029 extends JApplet {
-    private boolean res;
+public class bug4150029 {
+    private static JFrame frame;
+    private static JFileChooser fileChooser;
+    private static Robot robot;
+    private static File testDir;
+    private static File subDir;
```
Comment (kumarabhi006) [inline test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java:52]: `testDir` can be declared as local variable.
Probably `subDir` also can be moved as local variable provided you create JFileChooser instance with default directory and then set current directory inside main method after createAndShowUI method call.

something like this..

`SwingUtilities.invokeAndWait(() -> {
                createAndShowUI();
                fileChooser.setCurrentDirectory(subDir);
            });`

File: test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java [hunk 12]
```diff
@@ -21,73 +21,119 @@
...
-        }
+        // check backspace key at subDir level
+        clickBackSpace();
 
-        try {
-            subDir.delete();
-            testDir.delete();
-        } catch (SecurityException e) {
-            e.printStackTrace();
+        if (!(prevDir != crntDir)) {
```
Comment (kumarabhi006) [inline test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java:128]: simplified condition - 
```suggestion
        if (prevDir == crntDir) {
```

File: test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java [hunk 13]
```diff
@@ -21,73 +21,119 @@
...
+        frame.pack();
+        frame.setLocationRelativeTo(null);
+        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
+        frame.setVisible(true);
+    }
 
-        JFileChooser fileChooser = new JFileChooser(testDir);
+    private static void doTesting() throws Exception {
+        SwingUtilities.invokeAndWait(() -> {
+            p = frame.getLocationOnScreen();
```
Comment (kumarabhi006) [inline test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java:117]: May remove extra blank line.

**Review by kumarabhi006 — APPROVED** [review]
Looks good now. Only point is "should we move this test outside folder as html file is deleted now?".
Hope you have tested in CI for all platform.

**Comment by alisenchung** [comment]
> Looks good now. Only point is "should we move this test outside folder as html file is deleted now?". Hope you have tested in CI for all platform.

Clientlibs is green, also I'll keep this inside the folder for bookkeeping

**Comment by kumarabhi006** [comment]
> /integrate

You need a "Reviewer" approval to integrate.

**Review by DamonGuy — APPROVED** [review]
Agreed. Looks good now, especially if CI is green. Still need "R"eviewer approval.

File: test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java [hunk 14]
```diff
@@ -21,73 +21,117 @@
...
-        }
+        // check backspace key at subDir level
+        clickBackSpace();
 
-        try {
-            subDir.delete();
-            testDir.delete();
-        } catch (SecurityException e) {
-            e.printStackTrace();
+        if (prevDir == crntDir) {
```
Comment (honkar-jdk) [inline test/jdk/javax/swing/JFileChooser/4150029/bug4150029.java:126]: Wrong comparison operator should be .equals(). 
You might want to run the CI testing again with the updated changes just to make sure it works the same on all platforms.

```suggestion
        if (prevDir.equals(crntDir)) {
```

**Review by honkar-jdk — CHANGES_REQUESTED** [review]

**Review by honkar-jdk — APPROVED** [review]
LGTM

**Pull request closed without merging**